plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/panel/GitPanelPresenter.java

```diff
@@ -0,0 +1,360 @@
+import org.eclipse.che.ide.ext.git.client.GitServiceClient;
+import org.eclipse.che.ide.ext.git.client.compare.AlteredFiles;
+import org.eclipse.che.ide.ext.git.client.compare.FileStatus.Status;
+import org.eclipse.che.ide.ext.git.client.compare.MutableAlteredFiles;
+import org.eclipse.che.ide.ext.git.client.compare.changespanel.ChangesPanelPresenter;
+import org.vectomatic.dom.svg.ui.SVGResource;
+
+/**
+ * Presenter for the Git panel.
+ *
+ * @author Mykola Morhun
+ */
+@Singleton
+public class GitPanelPresenter extends BasePresenter
+    implements GitPanelView.ActionDelegate, GitEventsSubscriber, ResourceChangedHandler {
+
+  private static final String VCS_GIT = "git";
+  private static final String REVISION = "HEAD";
+
+  private final GitPanelView view;
+  private final GitServiceClient service;
+  private final ChangesPanelPresenter changesPanelPresenter;
+  private final AppContext appContext;
+  private final EventBus eventBus;
+  private final NotificationManager notificationManager;
+  private final GitEventSubscribable subscribeToGitEvents;
+  private final GitResources gitResources;
+  private final GitLocalizationConstant locale;
+
+  private boolean initialized;
+  private Map<String, MutableAlteredFiles> changes; // project name -> changes
+  private String selectedProjectName;
+
+  @Inject
+  public GitPanelPresenter(
+      GitPanelView view,
+      GitServiceClient service,
+      ChangesPanelPresenter changesPanelPresenter,
+      WorkspaceAgent workspaceAgent,
+      AppContext appContext,
+      EventBus eventBus,
+      GitEventSubscribable subscribeToGitEvents,
+      NotificationManager notificationManager,
+      GitResources gitResources,
+      GitLocalizationConstant locale) {
+    this.view = view;
+    this.service = service;
+    this.changesPanelPresenter = changesPanelPresenter;
+    this.appContext = appContext;
+    this.eventBus = eventBus;
+    this.subscribeToGitEvents = subscribeToGitEvents;
+    this.notificationManager = notificationManager;
+    this.gitResources = gitResources;
+    this.locale = locale;
+
+    this.view.setDelegate(this);
+    this.view.setChangesPanelView(this.changesPanelPresenter.getView());
+
+    if (partStack == null || !partStack.containsPart(this)) {
+      workspaceAgent.openPart(this, PartStackType.NAVIGATION);
+    }
+
+    this.initialized = false;
+  }
+
+  private void registerEventHandlers() {
+    eventBus.addHandler(ResourceChangedEvent.getType(), this);
+
+    subscribeToGitEvents.addSubscriber(this);
+  }
+
+  /** Invoked each time when panel is activated. */
+  @Override
+  public void onOpen() {
+    if (!initialized) {
+      loadPanelData();
+      registerEventHandlers();
+
+      initialized = true;
+    }
+  }
+
+  /** Queries from server all data needed to initialize the panel. */
+  private void loadPanelData() {
+    this.changes = new HashMap<>();
+
+    for (Project project : appContext.getProjects()) {
```

Comment:
How about streams?

Reply:
It's possible, but to me stream here look a bit awkward because of pure java array conversion to stream like: `Arrays.stream(appContext.getProjects()).filter(...`